### PR TITLE
Fix Abbreviation for Escaped Ampersand

### LINKED ISSUE #8948: Improve Equality check for Journal abbreviations
I have a question regarding the use of ampersands `&`, I often come across `\&` in my `.bib` files when an ampersand is used, JabRef doesn't seem to recognize that journaltitle if the name is not exactly the same. Do we need to add entries for all occurences of `&` and `\&` and `and` as well?

_Originally posted by @jhossbach in https://github.com/JabRef/abbrv.jabref.org/issues/100#issuecomment-1129177022_

### PR DESCRIPTION
**Pull request by ANUu7312578**
Fixes #8948.

This pull request attempts to fix issue #8948. The main #8948 issue has 3 sub-components: 
1. Abbreviations should always be without \\& --> There is a CI check "necessary" at the abbrv repo
2. JabRef should be able to handle both & and \\& when abbrevating and unabbreviating
3. When putting a result, \\& should be used

The first sub-component, is currently being implemented by the pull request https://github.com/JabRef/abbrv.jabref.org/pull/108. The changes highlighted in this pull request attempt to fix the second sub-component. It seems that the third component has already been implemented except that it has to be manually enabled by users.

A test which attempt to verify that unescaped ampersands in journal titles actually result in abbreviations has also been added.

Any feedback would be appreciated.

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/main/java/org/jabref/logic/journals/JournalAbbreviationRepository.java [hunk 1]
```diff
@@ -48,7 +49,8 @@ private static boolean isMatchedAbbreviated(String name, Abbreviation abbreviati
      * Letters) or its abbreviated form (e.g. Phys. Rev. Lett.).
      */
     public boolean isKnownName(String journalName) {
-        String journal = journalName.trim();
+        // Trims the String and also replaces any instances of "\&" with "&" for abbreviation search purposes
```
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/JournalAbbreviationRepository.java:52]: `Trims the String and also replaces any instances of "\&" with "&"` is a literal explanation of the code and can be removed.

```suggestion
        // The journal lists contains unescaped  only, thus "\&" is replaced by "&" for abbreviation search purposes
```

File: src/main/java/org/jabref/logic/journals/JournalAbbreviationRepository.java [hunk 2]
```diff
@@ -75,7 +77,8 @@ public boolean isAbbreviatedName(String journalName) {
      * @param input The journal name (either abbreviated or full name).
      */
     public Optional<Abbreviation> get(String input) {
-        String journal = input.trim();
+        // Trims the String and also replaces any instances of "\&" with "&" for abbreviation search purposes
+        String journal = input.trim().replaceAll(Matcher.quoteReplacement("\\&"), "&");
```
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/JournalAbbreviationRepository.java:81]: See above 😇

**Review by koppor — CHANGES_REQUESTED** [review]
Minor comment, otherwise: Looks good to me.

The latexfreefield could not be used for checking the journal name? (`org.jabref.model.entry.BibEntry#getResolvedFieldOrAliasLatexFree`)

**Comment by ANUu7312578** [comment]
@koppor I must admit that I did not understand exactly what `org.jabref.model.entry.BibEntry#getResolvedFieldOrAliasLatexFree` was doing or what its parameters meant. So, I just used `replaceAll` because it seemed easier :sweat_smile:. I have pushed a commit which attempts to use `org.jabref.model.entry.BibEntry#getResolvedFieldOrAliasLatexFree`. Could you provide some feedback? From my testing, it seems to accomplish the same job that `replaceAll` does. Thanks!

File: src/main/java/org/jabref/logic/journals/JournalAbbreviationRepository.java (hunk 1 above)
Comment (ANUu7312578) [reply src/main/java/org/jabref/logic/journals/JournalAbbreviationRepository.java]: Removed!

File: src/main/java/org/jabref/logic/journals/JournalAbbreviationRepository.java (hunk 2 above)
Comment (ANUu7312578) [reply src/main/java/org/jabref/logic/journals/JournalAbbreviationRepository.java]: Removed!

**Comment by Siedlerchr** [comment]
@ANUu7312578  JabRef supports both  BibTeX and biblatex. The old BibTeX format does not understand unicode and also special characters like & needs to be escaped using the special latex syntax.
The method GetResolvedFieldOrAlias does a few more things but the relevant part here is that it returns a latex command free value of the field (so basically the unicode)

**Comment by koppor** [comment]
I miss test cases for following BibTeX entry:

```bibtex
@article{test,
  journal = {ACS Applied Materials \\& Interfaces},
}
```

The result should be


```bibtex
@article{test,
  journal = {ACS Appl. Mater. Interfaces},
}
```

Second test:

```bibtex
@article{test,
  journal = {ACS Applied Materials \\& Interfaces},
}
```

The result should be


```bibtex
@article{test,
  journal = {ACS Appl. Mater. Interfaces},
}
```


Maybe, there are these tests, but better one more than no one.

**Comment by ANUu7312578** [comment]
@koppor Hi! Sorry, but I'm not sure what you mean when you say that you miss test cases for the BibTeX entries. Do you mean that I should create tests which actually use a `BibEntry` object which is effectively

```
@article{test,
  journal = {ACS Applied Materials \\& Interfaces},
}
```

and try to abbreviate the `BibEntry` object journal title in order to test if it abbreviates into

```
@article{test,
  journal = {ACS Appl. Mater. Interfaces},
}
```

If so, I am not actually sure which function(s) to call to abbreviate the journal name in a `BibEntry` object. Would you be able to provide some pointers on this? It seems that `org.jabref.gui.journals.UndoableAbbreviator` is the only class which calls `org.jabref.logic.journals.JournalAbbreviationRepository`. However, this class seems to be more for the GUI and not for internal logic. If not, could you explain a bit more what you meant? Any clarification would be welcome!

Additionally, I am not sure I see the difference between the "first test" and the "second test". It seems that they both have the journal name as "ACS Applied Materials \\\\& Interfaces" and the abbreviated name as "ACS Appl. Mater. Interfaces". Also, I was wondering why the abbreviation of "ACS Applied Materials \\\\& Interfaces" would be "ACS Appl. Mater. Interfaces". Am I mistaken in thinking that the journal name should only abbreviate if the ampersand in the journal name is unescaped or only escaped once (e.g., "ACS Applied Materials \\& Interfaces" -> "ACS Appl. Mater. Interfaces")?

Sorry for all the questions!

**Comment by Siedlerchr** [comment]
Hi, 
yes it's a bit confusing. But I think of a kind of integration test here:
You would use the logic classes:
UndoableUnabbreviator and UndoableAbbreviator 

Create a new BibEntry and a new databasecontext

**Comment by koppor** [comment]
> @koppor Hi! Sorry, but I'm not sure what you mean when you say that you miss test cases for the BibTeX entries. Do you mean that I should create tests which actually use a `BibEntry` object 

Yes.

> Additionally, I am not sure I see the difference between the "first test" and the "second test". 

I updated the second test to have not any escapings.

> Am I mistaken in thinking that the journal name should only abbreviate if the ampersand in the journal name is unescaped or only escaped once (e.g., "ACS Applied Materials \& Interfaces" -> "ACS Appl. Mater. Interfaces")?

The thing is, it should be able to abbreviate in **both** cases. This should be shown by a test.

**Comment by ANUu7312578** [comment]
Hi! I've attempted to write some tests according to the test cases specified by @koppor and using `UndoableUnabbreviator` and `UndoableAbbreviator` as suggested by @Siedlerchr. Currently, these test cases reside in `org.jabref.logic.bibtex.BibEntryWriterTest`. However, this causes the `restrictUsagesInLogic` test to fail because a `CompoundEdit` object is required to abbreviate a journal title. Could I get some feedback as to whether the tests I wrote are what was actually wanted and if I should move these tests to another test class in order to stop `restrictUsagesInLogic` from failing? Or, if there is another way to resolve `restrictUsagesInLogic`, any advice would be appreciated. Thanks!

**Comment by koppor** [comment]
The tests test the Journal Abbreviations. Thus, they should be in the test class for the Journal Abbreviator.

An Exception to the ArchUnit rules can be added in this case. There needs to be some refinement so that one can abbreviate without using a gui class.

I was thinking, there is a method
In: BibEntry
out: BibEntry abbreviated

Seems that this resides in gui only?

**Comment by Siedlerchr** [comment]
@koppor  The logic methods are in UndoableAbbreviator and UndoableUnabbreviator. They take an entry, a field and a databasecontext  However, they require the swing compound edit for undoing

**Comment by ANUu7312578** [comment]
I have moved the tests from the `org.jabref.logic.bibtex.BibEntryWriterTest` class to the `org.jabref.logic.journals.JournalAbbreviationRepositoryTest` class like @koppor suggested. Hopefully it is now in the appropriate test class.

Is there anything else which needs fixing? I am also not really sure how to add an exception to the ArchUnit rules 😅.

**Comment by Siedlerchr** [comment]
@ANUu7312578  For the test I would adapt this as follows:
Then you should be able to annotate the test class with `@AllowedToUseSwing`
 ```
   @ArchTest
    public static void restrictUsagesInLogic(JavaClasses classes) {
        noClasses().that().resideInAPackage(PACKAGE_ORG_JABREF_LOGIC)
                    .and().areNotAnnotatedWith(AllowedToUseSwing.class)
                   .should().dependOnClassesThat().resideInAPackage(PACKAGE_JAVAX_SWING)
                   .orShould().dependOnClassesThat().haveFullyQualifiedName(CLASS_ORG_JABREF_GLOBALS)
                   .check(classes);
    }
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -74,6 +74,7 @@ Note that this project **does not** adhere to [Semantic Versioning](http://semve
 - We fixed an issue where hitting enter on the search field within the preferences dialog closed the dialog. [koppor#630](https://github.com/koppor/jabref/issues/630)
 - We fixed a typo within a connection error message. [koppor#625](https://github.com/koppor/jabref/issues/625)
 - We fixed an issue where the 'close dialog' key binding was not closing the Preferences dialog. [#8888](https://github.com/jabref/jabref/issues/8888)
+- We fixed an issue where journal abbreviations would not abbreviate journal titles with escaped ampersands (\\&) which are written in accordance with Latex format. [#8948](https://github.com/JabRef/jabref/issues/8948)
```
Comment (koppor) [inline CHANGELOG.md:77]: I would remove the last part, that information should be made available in the internet elsewhere 😃

```suggestion
- We fixed an issue where journal abbreviations would not abbreviate journal titles with escaped ampersands (\\&). [#8948](https://github.com/JabRef/jabref/issues/8948)
```

File: src/test/java/org/jabref/logic/journals/JournalAbbreviationRepositoryTest.java [hunk 4]
```diff
@@ -134,4 +156,97 @@ void getFromFullName() {
...
+        String expectedAbbreviatedJournalEntry = """
+                @Article{,
+                  journal = {ACS Appl. Mater. Interfaces},
+                }
+                """.replaceAll("\n", OS.NEWLINE);
+        // @formatter:on
+
+        undoableAbbreviator.abbreviate(bibDatabase, entryWithEscapedAmpersandInJournal, StandardField.JOURNAL, new CompoundEdit());
+        bibEntryWriter.write(entryWithEscapedAmpersandInJournal, bibWriter, BibDatabaseMode.BIBLATEX);
+        assertEquals(expectedAbbreviatedJournalEntry, stringWriter.toString());
```
Comment (koppor) [inline src/test/java/org/jabref/logic/journals/JournalAbbreviationRepositoryTest.java:196]: Comparison on a String bases should only be used for serialization tests.

In your case, do

```suggestion
        undoableAbbreviator.abbreviate(bibDatabase, entryWithEscapedAmpersandInJournal, StandardField.JOURNAL, new CompoundEdit());
        BibEntry expectedAbbreviatedJournalEntry = new BibEntry(StandardEntryType.Article)
                .withField(StandardField.JOURNAL, "ACS Appl. Mater. Interfaces");
        assertEquals(expectedAbbreviatedJournalEntry , entryWithEscapedAmpersandInJournal);
```

**Review by koppor — CHANGES_REQUESTED** [review]
Thank you for the tests.

They are currently too complicated. I made another proposal

**Comment by ANUu7312578** [comment]
Thanks to @Siedlerchr for the advice on how to add an exception to the ArchUnit rules! Hopefully, I've now accomplished everything @koppor asked for. The `restrictUsagesInLogic` test should no longer be failing. Please let me know if there is anything else that I should do! 😃

**Comment by Siedlerchr** [comment]
Thanks again for the quick follow up! From my side it looks already good from my side

File: src/main/java/org/jabref/logic/journals/JournalAbbreviationRepository.java [hunk 5]
```diff
@@ -48,7 +49,7 @@ private static boolean isMatchedAbbreviated(String name, Abbreviation abbreviati
      * Letters) or its abbreviated form (e.g. Phys. Rev. Lett.).
      */
     public boolean isKnownName(String journalName) {
-        String journal = journalName.trim();
+        String journal = journalName.trim().replaceAll(Matcher.quoteReplacement("\\&"), "&");
```
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/JournalAbbreviationRepository.java:52]: I think, there is the possibility to pre-compile the Regex. Was it with the `Pattern` class?

File: src/test/java/org/jabref/logic/journals/JournalAbbreviationRepositoryTest.java [hunk 6]
```diff
@@ -134,4 +146,61 @@ void getFromFullName() {
...
+    @Test
+    void testAbbreviationsWithEscapedAmpersand() {
+        assertEquals(new Abbreviation("ACS Applied Materials & Interfaces", "ACS Appl. Mater. Interfaces"), repository.get("ACS Applied Materials & Interfaces").get());
+        assertEquals(new Abbreviation("ACS Applied Materials & Interfaces", "ACS Appl. Mater. Interfaces"), repository.get("ACS Applied Materials \\& Interfaces").get());
+        assertEquals(new Abbreviation("Antioxidants & Redox Signaling", "Antioxid. Redox Signaling"), repository.get("Antioxidants & Redox Signaling").get());
+        assertEquals(new Abbreviation("Antioxidants & Redox Signaling", "Antioxid. Redox Signaling"), repository.get("Antioxidants \\& Redox Signaling").get());
+
+        repository.addCustomAbbreviation(new Abbreviation("Long & Name", "L. N.", "LN"));
+        assertEquals(new Abbreviation("Long & Name", "L. N.", "LN"), repository.get("Long & Name").get());
+        assertEquals(new Abbreviation("Long & Name", "L. N.", "LN"), repository.get("Long \\& Name").get());
```
Comment (koppor) [inline src/test/java/org/jabref/logic/journals/JournalAbbreviationRepositoryTest.java:159]: Normally, we expect one assertation per test case. Thus, this method should be split into several ones. `@ParamterizedTest` is the the thign to use.

File: src/test/java/org/jabref/logic/journals/JournalAbbreviationRepositoryTest.java [hunk 7]
```diff
@@ -134,4 +146,61 @@ void getFromFullName() {
...
+    }
+
+    @Test
+    void testJournalAbbreviationWithEscapedAmpersand() {
+        BibDatabase bibDatabase = new BibDatabase();
+        JournalAbbreviationRepository journalAbbreviationRepository = JournalAbbreviationLoader.loadBuiltInRepository();
+        UndoableAbbreviator undoableAbbreviator = new UndoableAbbreviator(journalAbbreviationRepository, AbbreviationType.DEFAULT);
+
+        BibEntry entryWithEscapedAmpersandInJournal = new BibEntry(StandardEntryType.Article);
+        entryWithEscapedAmpersandInJournal.setField(StandardField.JOURNAL, "ACS Applied Materials \\& Interfaces");
```
Comment (koppor) [inline src/test/java/org/jabref/logic/journals/JournalAbbreviationRepositoryTest.java:169]:
```suggestion
        BibEntry entryWithEscapedAmpersandInJournal = new BibEntry(StandardEntryType.Article)
            .withField(StandardField.JOURNAL, "ACS Applied Materials \\& Interfaces");
```

**Review by koppor — APPROVED** [review]
Thank you for keeping working on this.

I added some comments 😇. - Since they are more general hints than really required, I go forward with a merge 😅. You can do a follow-up PR if you have time ^^

**Pull request merged**